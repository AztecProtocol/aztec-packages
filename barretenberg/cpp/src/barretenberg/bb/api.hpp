#pragma once
#include <filesystem>

namespace bb {

class API {
  public:
    struct Flags {
        std::optional<std::string> output_type; // bytes, fields, bytes_and_fields, fields_msgpack
        std::optional<std::string> input_type;  // compiletime_stack, runtime_stack
    };

    virtual void prove(const Flags& flags,
                       const std::filesystem::path& bytecode_path,
                       const std::filesystem::path& witness_path,
                       const std::filesystem::path& output_dir) = 0;

    virtual bool verify(const Flags& flags,
                        const std::filesystem::path& proof_path,
                        const std::filesystem::path& vk_path) = 0;

    virtual bool prove_and_verify(const Flags& flags,
                                  const std::filesystem::path& bytecode_path,
                                  const std::filesystem::path& witness_path) = 0;

    virtual void gates(const Flags& flags,
                       const std::filesystem::path& bytecode_path,
                       const std::filesystem::path& witness_path) = 0;

    virtual void contract(const Flags& flags,
                          const std::filesystem::path& output_path,
                          const std::filesystem::path& vk_path) = 0;

    virtual void to_fields(const Flags& flags,
                           const std::filesystem::path& proof_path,
                           const std::filesystem::path& vk_path,
                           const std::filesystem::path& output_path) = 0;

    virtual void write_arbitrary_valid_proof_and_vk_to_file(const API::Flags& flags,
                                                            const std::filesystem::path& output_dir) = 0;
};
} // namespace bb
