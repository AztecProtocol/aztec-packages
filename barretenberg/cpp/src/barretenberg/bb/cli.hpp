#pragma once

namespace bb {
int parse_and_run_cli_command(int argc, char* argv[]);
}
