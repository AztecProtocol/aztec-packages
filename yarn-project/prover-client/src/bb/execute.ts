import { sha256 } from '@aztec/foundation/crypto';
import { type LogFn } from '@aztec/foundation/log';
import { Timer } from '@aztec/foundation/timer';
import { type NoirCompiledCircuit } from '@aztec/types/noir';

import * as proc from 'child_process';
import * as fs from 'fs/promises';

import { BBNativeRollupProver, type BBProverConfig } from '../prover/bb_prover.js';

export enum BB_RESULT {
  SUCCESS,
  FAILURE,
  ALREADY_PRESENT,
}

export type BBSuccess = {
  status: BB_RESULT.SUCCESS | BB_RESULT.ALREADY_PRESENT;
  duration: number;
  path?: string;
};

export type BBFailure = {
  status: BB_RESULT.FAILURE;
  reason: string;
};

export type BBResult = BBSuccess | BBFailure;

/**
 * Invokes the Barretenberg binary with the provided command and args
 * @param pathToBB - The path to the BB binary
 * @param command - The command to execute
 * @param args - The arguments to pass
 * @param logger - A log function
 * @param resultParser - An optional handler for detecting success or failure
 * @returns The completed partial witness outputted from the circuit
 */
export function executeBB(
  pathToBB: string,
  command: string,
  args: string[],
  logger: LogFn,
  resultParser = (code: number) => code === 0,
) {
  return new Promise<BB_RESULT.SUCCESS | BB_RESULT.FAILURE>((resolve, reject) => {
    // spawn the bb process
    const acvm = proc.spawn(pathToBB, [command, ...args]);
    acvm.stdout.on('data', data => {
      const message = data.toString('utf-8').replace(/\n$/, '');
      logger(message);
    });
    acvm.stderr.on('data', data => {
      const message = data.toString('utf-8').replace(/\n$/, '');
      logger(message);
    });
    acvm.on('close', (code: number) => {
      if (resultParser(code)) {
        resolve(BB_RESULT.SUCCESS);
      } else {
        reject();
      }
    });
  }).catch(_ => BB_RESULT.FAILURE);
}

const bytecodeHashFilename = 'bytecode_hash';
const bytecodeFilename = 'bytecode';
const proofFileName = 'proof';

/**
 * Used for generating either a proving or verification key, will exit early if the key already exists
 * It assumes the provided working directory is one where the caller wishes to maintain a permanent set of keys
 * It is not considered a temporary directory
 * @param pathToBB - The full path to the bb binary
 * @param workingDirectory - The directory into which the key should be created
 * @param circuitName - An identifier for the circuit
 * @param compiledCircuit - The compiled circuit
 * @param key - The type of key, either 'pk' or 'vk'
 * @param log - A logging function
 * @param force - Force the key to be regenerated even if it already exists
 * @returns An instance of BBResult
 */
export async function generateKeyForNoirCircuit(
  pathToBB: string,
  workingDirectory: string,
  circuitName: string,
  compiledCircuit: NoirCompiledCircuit,
  key: 'vk' | 'pk',
  log: LogFn,
  force = false,
): Promise<BBSuccess | BBFailure> {
  const bytecode = Buffer.from(compiledCircuit.bytecode, 'base64');

  // The key generation is written to e.g. /workingDirectory/pk/BaseParityArtifact/pk
  // The bytecode hash file is also written here as /workingDirectory/pk/BaseParityArtifact/bytecode-hash
  // The bytecode is written to e.g. /workingDirectory/pk/BaseParityArtifact/bytecode
  // The bytecode is removed after the key is generated, leaving just the hash file
  const circuitOutputDirectory = `${workingDirectory}/${key}/${circuitName}`;
  const bytecodeHashPath = `${circuitOutputDirectory}/${bytecodeHashFilename}`;
  const bytecodePath = `${circuitOutputDirectory}/${bytecodeFilename}`;
  const bytecodeHash = sha256(bytecode);

  const outputPath = `${circuitOutputDirectory}/${key}`;

  // ensure the directory exists
  await fs.mkdir(circuitOutputDirectory, { recursive: true });

  // Generate the key if we have been told to, or there is no bytecode hash
  let mustRegenerate =
    force ||
    (await fs
      .access(bytecodeHashPath, fs.constants.R_OK)
      .then(_ => false)
      .catch(_ => true));

  if (!mustRegenerate) {
    // Check to see if the bytecode hash has changed from the stored value
    const data: Buffer = await fs.readFile(bytecodeHashPath).catch(_ => Buffer.alloc(0));
    mustRegenerate = data.length == 0 || !data.equals(bytecodeHash);
  }

  if (!mustRegenerate) {
    // No need to generate, early out
    return { status: BB_RESULT.ALREADY_PRESENT, duration: 0, path: outputPath };
  }

  // Check we have access to bb
  const binaryPresent = await fs
    .access(pathToBB, fs.constants.R_OK)
    .then(_ => true)
    .catch(_ => false);
  if (!binaryPresent) {
    return { status: BB_RESULT.FAILURE, reason: `Failed to find bb binary at ${pathToBB}` };
  }

  // We are now going to generate the key
  try {
    // Write the bytecode to the working directory
    await fs.writeFile(bytecodePath, bytecode);

    // args are the output path and the input bytecode path
    const args = ['-o', outputPath, '-b', bytecodePath];
    const timer = new Timer();
    const result = await executeBB(pathToBB, `write_${key}`, args, log);
    const duration = timer.ms();
    // Cleanup the bytecode file
    await fs.rm(bytecodePath, { force: true });
    if (result == BB_RESULT.SUCCESS) {
      // Store the bytecode hash so we don't need to regenerate at a later time
      await fs.writeFile(bytecodeHashPath, bytecodeHash);
      return { status: BB_RESULT.SUCCESS, duration, path: outputPath };
    }
    // Not a great error message here but it is difficult to decipher what comes from bb
    return { status: BB_RESULT.FAILURE, reason: `Failed to generate key` };
  } catch (error) {
    return { status: BB_RESULT.FAILURE, reason: `${error}` };
  }
}

/**
 * Used for generating proofs of noir circuits.
 * It is assumed that the working directory is a temporary and/or random directory used solely for generating this proof.
 * @param pathToBB - The full path to the bb binary
 * @param workingDirectory - A working directory for use by bb
 * @param circuitName - An identifier for the circuit
 * @param compiledCircuit - The compiled circuit
 * @param inputWitnessFile - The circuit input witness
 * @param log - A logging function
 * @returns An object containing a result indication, the location of the proof and the duration taken
 */
export async function generateProof(
  pathToBB: string,
  workingDirectory: string,
  circuitName: string,
  compiledCircuit: NoirCompiledCircuit,
  inputWitnessFile: string,
  log: LogFn,
): Promise<BBFailure | BBSuccess> {
  // Check that the working directory exists
  try {
    await fs.access(workingDirectory);
  } catch (error) {
    return { status: BB_RESULT.FAILURE, reason: `Working directory ${workingDirectory} does not exist` };
  }

  // The bytecode is written to e.g. /workingDirectory/BaseParityArtifact-bytecode
  const bytecodePath = `${workingDirectory}/${circuitName}-bytecode`;
  const bytecode = Buffer.from(compiledCircuit.bytecode, 'base64');

  // The proof is written to e.g. /workingDirectory/proof
  const outputPath = `${workingDirectory}/${proofFileName}`;

  const binaryPresent = await fs
    .access(pathToBB, fs.constants.R_OK)
    .then(_ => true)
    .catch(_ => false);
  if (!binaryPresent) {
    return { status: BB_RESULT.FAILURE, reason: `Failed to find bb binary at ${pathToBB}` };
  }

  try {
    // Write the bytecode to the working directory
    await fs.writeFile(bytecodePath, bytecode);
    const args = ['-o', outputPath, '-b', bytecodePath, '-w', inputWitnessFile];
    const command = 'prove';
    const timer = new Timer();
    const logFunction = (message: string) => {
      log(`${circuitName} BB out - ${message}`);
    };
    const result = await executeBB(pathToBB, command, args, logFunction);
    const duration = timer.ms();
    // cleanup the bytecode
    await fs.rm(bytecodePath, { force: true });
    if (result == BB_RESULT.SUCCESS) {
      return { status: BB_RESULT.SUCCESS, duration, path: outputPath };
    }
    // Not a great error message here but it is difficult to decipher what comes from bb
    return { status: BB_RESULT.FAILURE, reason: `Failed to generate proof` };
  } catch (error) {
    return { status: BB_RESULT.FAILURE, reason: `${error}` };
  }
}

/**
 * Used for verifying proofs of noir circuits
 * @param pathToBB - The full path to the bb binary
 * @param proofFullPath - The full path to the proof to be verified
 * @param verificationKeyPath - The full path to the circuit verification key
 * @param log - A logging function
 * @returns An object containing a result indication and duration taken
 */
export async function verifyProof(
  pathToBB: string,
  proofFullPath: string,
  verificationKeyPath: string,
  log: LogFn,
): Promise<BBFailure | BBSuccess> {
  const binaryPresent = await fs
    .access(pathToBB, fs.constants.R_OK)
    .then(_ => true)
    .catch(_ => false);
  if (!binaryPresent) {
    return { status: BB_RESULT.FAILURE, reason: `Failed to find bb binary at ${pathToBB}` };
  }

  try {
    const args = ['-p', proofFullPath, '-k', verificationKeyPath];
    const timer = new Timer();
    const result = await executeBB(pathToBB, 'verify', args, log);
    const duration = timer.ms();
    if (result == BB_RESULT.SUCCESS) {
      return { status: BB_RESULT.SUCCESS, duration };
    }
    // Not a great error message here but it is difficult to decipher what comes from bb
    return { status: BB_RESULT.FAILURE, reason: `Failed to verify proof` };
  } catch (error) {
    return { status: BB_RESULT.FAILURE, reason: `${error}` };
  }
}

/**
 * Used for generating all verification keys required by server protocol circuits
 * @param pathToBB - The full path to the bb binary
 * @param workingDirectory - The directory to be used for the keys
 * @param log - A logging function
 */
export async function generateAllServerVks(pathToBB: string, workingDirectory: string, log: LogFn) {
  const bbConfig: BBProverConfig = {
    bbBinaryPath: pathToBB,
    bbWorkingDirectory: workingDirectory,

    // These aren't needed for this
    acvmBinaryPath: '',
    acvmWorkingDirectory: '',
    circuitFilter: [],
  };
  // This will generate all of the server circuit verification keys for us
  try {
    await BBNativeRollupProver.generateVerificationKeys(bbConfig);
  } catch (error) {
    log(`Failed to generate verification keys: ${error}`);
  }
}
