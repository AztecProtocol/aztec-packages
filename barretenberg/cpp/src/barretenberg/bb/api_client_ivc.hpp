#pragma once

#include "barretenberg/bb/acir_format_getters.hpp"
#include "barretenberg/bb/api.hpp"
#include "barretenberg/bb/init_srs.hpp"
#include "barretenberg/common/throw_or_abort.hpp"
#include "libdeflate.h"

namespace bb {

template <typename T> std::shared_ptr<T> read_to_shared_ptr(const std::filesystem::path& path)
{
    return std::make_shared<T>(from_buffer<T>(read_file(path)));
};

// TODO(#7371): this could probably be more idiomatic
template <typename T> T unpack_from_file(const std::filesystem::path& filename)
{
    std::ifstream fin;
    fin.open(filename, std::ios::ate | std::ios::binary);
    if (!fin.is_open()) {
        throw std::invalid_argument("file not found");
    }
    if (fin.tellg() == -1) {
        throw std::invalid_argument("something went wrong");
    }

    uint64_t fsize = static_cast<uint64_t>(fin.tellg());
    fin.seekg(0, std::ios_base::beg);

    T result;
    char* encoded_data = new char[fsize];
    fin.read(encoded_data, static_cast<std::streamsize>(fsize));
    msgpack::unpack(encoded_data, fsize).get().convert(result);
    return result;
}

// TODO(#7371) find a home for this
acir_format::WitnessVector witness_map_to_witness_vector(std::map<std::string, std::string> const& witness_map)
{
    acir_format::WitnessVector wv;
    size_t index = 0;
    for (auto& e : witness_map) {
        uint64_t value = std::stoull(e.first);
        // ACIR uses a sparse format for WitnessMap where unused witness indices may be left unassigned.
        // To ensure that witnesses sit at the correct indices in the `WitnessVector`, we fill any indices
        // which do not exist within the `WitnessMap` with the dummy value of zero.
        while (index < value) {
            wv.push_back(fr(0));
            index++;
        }
        wv.push_back(fr(uint256_t(e.second)));
        index++;
    }
    return wv;
}

std::vector<uint8_t> decompress(uint8_t* bytes, size_t size)
{
    std::vector<uint8_t> content;
    // initial size guess
    content.resize(1024ULL * 128ULL);
    for (;;) {
        auto decompressor = std::unique_ptr<libdeflate_decompressor, void (*)(libdeflate_decompressor*)>{
            libdeflate_alloc_decompressor(), libdeflate_free_decompressor
        };
        size_t actual_size = 0;
        libdeflate_result decompress_result = libdeflate_gzip_decompress(
            decompressor.get(), bytes, size, std::data(content), std::size(content), &actual_size);
        if (decompress_result == LIBDEFLATE_INSUFFICIENT_SPACE) {
            // need a bigger buffer
            content.resize(content.size() * 2);
            continue;
        }
        if (decompress_result == LIBDEFLATE_BAD_DATA) {
            throw std::invalid_argument("bad gzip data in bb main");
        }
        content.resize(actual_size);
        break;
    }
    return content;
}

class ClientIVCAPI : public API {
    static std::vector<acir_format::AcirProgram> _build_folding_stack(const std::string& input_type,
                                                                      const std::filesystem::path& bytecode_path,
                                                                      const std::filesystem::path& witness_path)
    {
        using namespace acir_format;

        std::vector<AcirProgram> folding_stack;

        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1162): Efficiently unify ACIR stack parsing
        if (input_type == "compiletime_stack") {
            auto program_stack =
                acir_format::get_acir_program_stack(bytecode_path, witness_path, /*honk_recursion=*/false);
            // Accumulate the entire program stack into the IVC
            while (!program_stack.empty()) {
                auto stack_item = program_stack.back();
                folding_stack.push_back(AcirProgram{ stack_item.constraints, stack_item.witness });
                program_stack.pop_back();
            }
        }

        if (input_type == "runtime_stack") {
            std::vector<std::string> gzipped_bincodes;
            std::vector<std::string> witness_data;
            gzipped_bincodes = unpack_from_file<std::vector<std::string>>(bytecode_path);
            witness_data = unpack_from_file<std::vector<std::string>>(witness_path);
            for (auto [bincode, wit] : zip_view(gzipped_bincodes, witness_data)) {
                // TODO(#7371) there is a lot of copying going on in bincode, we should make sure this writes as a
                // buffer in the future
                std::vector<uint8_t> constraint_buf =
                    decompress(reinterpret_cast<uint8_t*>(bincode.data()), bincode.size()); // NOLINT
                std::vector<uint8_t> witness_buf =
                    decompress(reinterpret_cast<uint8_t*>(wit.data()), wit.size()); // NOLINT

                AcirFormat constraints = circuit_buf_to_acir_format(constraint_buf, /*honk_recursion=*/false);
                WitnessVector witness = witness_buf_to_witness_data(witness_buf);

                folding_stack.push_back(AcirProgram{ constraints, witness });
            }
        }

        return folding_stack;
    };

    static ClientIVC _accumulate(std::vector<acir_format::AcirProgram>& folding_stack)
    {
        using Builder = MegaCircuitBuilder;
        using Program = acir_format::AcirProgram;

        using namespace acir_format;

        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1163) set these dynamically
        init_bn254_crs(1 << 20);
        init_grumpkin_crs(1 << 15);

        // TODO(#7371) dedupe this with the rest of the similar code
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1101): remove use of auto_verify_mode
        ClientIVC ivc{ { E2E_FULL_TEST_STRUCTURE }, /*auto_verify_mode=*/true };

        // Accumulate the entire program stack into the IVC
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1116): remove manual setting of is_kernel
        bool is_kernel = false;
        for (Program& program : folding_stack) {
            // Construct a bberg circuit from the acir representation then accumulate it into the IVC
            Builder circuit = acir_format::create_circuit<Builder>(
                program.constraints, true, 0, program.witness, false, ivc.goblin.op_queue);

            // Set the internal is_kernel flag based on the local mechanism only if it has not already been set to true
            if (!circuit.databus_propagation_data.is_kernel) {
                circuit.databus_propagation_data.is_kernel = is_kernel;
            }
            is_kernel = !is_kernel;

            // Do one step of ivc accumulator or, if there is only one circuit in the stack, prove that circuit. In this
            // case, no work is added to the Goblin opqueue, but VM proofs for trivials inputs are produced.
            ivc.accumulate(circuit, /*one_circuit=*/folding_stack.size() == 1);
        }

        return ivc;
    };

    static ClientIVC _accumulate_without_auto_verify(std::vector<acir_format::AcirProgram>& folding_stack)
    {
        using Builder = MegaCircuitBuilder;
        using Program = acir_format::AcirProgram;

        using namespace acir_format;

        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1163) set these dynamically
        init_bn254_crs(1 << 20);
        init_grumpkin_crs(1 << 15);

        // TODO(#7371) dedupe this with the rest of the similar code
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1101): remove use of auto_verify_mode
        ClientIVC ivc{ { E2E_FULL_TEST_STRUCTURE }, /*auto_verify_mode=*/false };

        // Accumulate the entire program stack into the IVC
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1116): remove manual setting of is_kernel once
        // databus has been integrated into noir kernel programs
        bool is_kernel = false;
        for (Program& program : folding_stack) {

            Builder circuit;

            is_kernel = !program.constraints.ivc_recursion_constraints.empty();
            if (is_kernel) {
                vinfo("Accumulating KERNEL.");
                circuit = create_kernel_circuit(program.constraints, ivc, program.witness);
            } else {
                vinfo("Accumulating APP.");
                circuit = create_circuit<Builder>(
                    program.constraints, /*recursive=*/false, 0, program.witness, false, ivc.goblin.op_queue);
            }

            // Do one step of ivc accumulator or, if there is only one circuit in the stack, prove that circuit. In this
            // case, no work is added to the Goblin opqueue, but VM proofs for trivial inputs are produced.
            ivc.accumulate(circuit, /*one_circuit=*/folding_stack.size() == 1);
        }

        return ivc;
    };

  public:
    void prove(const API::Flags& flags,
               const std::filesystem::path& bytecode_path,
               const std::filesystem::path& witness_path,
               const std::filesystem::path& output_dir) override
    {
        if (!flags.output_type || *flags.output_type != "fields_msgpack") {
            throw_or_abort("No output_type or output_type not supported");
        }

        if (!flags.input_type || !(*flags.input_type == "compiletime_stack" || *flags.input_type == "runtime_stack")) {
            throw_or_abort("No input_type or input_type not supported");
        }

        std::vector<acir_format::AcirProgram> folding_stack =
            _build_folding_stack(*flags.input_type, bytecode_path, witness_path);

        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1163) set these dynamically
        init_bn254_crs(1 << 20);
        init_grumpkin_crs(1 << 15);

        ClientIVC ivc;
        if (flags.no_auto_verify) {
            vinfo("performing accumulation WITHOUT auto-verify");
            ivc = _accumulate_without_auto_verify(folding_stack);
        } else {
            vinfo("performing accumulation with auto-verify");
            ivc = _accumulate(folding_stack);
        }
        ClientIVC::Proof proof = ivc.prove();

        // Write the proof and verification keys into the working directory in  'binary' format (in practice it seems
        // this directory is passed by bb.js)
        vinfo("writing ClientIVC proof and vk...");
        write_file(output_dir / "client_ivc_proof", to_buffer(proof));

        auto eccvm_vk = std::make_shared<ECCVMFlavor::VerificationKey>(ivc.goblin.get_eccvm_proving_key());
        auto translator_vk =
            std::make_shared<TranslatorFlavor::VerificationKey>(ivc.goblin.get_translator_proving_key());
        write_file(output_dir / "client_ivc_vk",
                   to_buffer(ClientIVC::VerificationKey{ ivc.honk_vk, eccvm_vk, translator_vk }));
    };

    /**
     * @brief Verifies a client ivc proof and writes the result to stdout
     *
     * Communication:
     * - proc_exit: A boolean value is returned indicating whether the proof is valid.
     *   an exit code of 0 will be returned for success and 1 for failure.
     *
     * @param proof_path Path to the file containing the serialized proof
     * @param vk_path Path to the serialized verification key of the final (MegaHonk) circuit in the stack
     * @param accumualtor_path Path to the file containing the serialized protogalaxy accumulator
     * @return true (resp., false) if the proof is valid (resp., invalid).
     */
    bool verify([[maybe_unused]] const API::Flags& flags,
                const std::filesystem::path& proof_path,
                const std::filesystem::path& vk_path) override
    {
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1163): Set these dynamically
        init_bn254_crs(1);
        init_grumpkin_crs(1 << 15);

        const auto proof = from_buffer<ClientIVC::Proof>(read_file(proof_path));
        const auto vk = from_buffer<ClientIVC::VerificationKey>(read_file(vk_path));

        vk.mega->pcs_verification_key = std::make_shared<VerifierCommitmentKey<curve::BN254>>();
        vk.eccvm->pcs_verification_key =
            std::make_shared<VerifierCommitmentKey<curve::Grumpkin>>(vk.eccvm->circuit_size + 1);
        vk.translator->pcs_verification_key = std::make_shared<VerifierCommitmentKey<curve::BN254>>();

        const bool verified = ClientIVC::verify(proof, vk);
        vinfo("verified: ", verified);
        return verified;
    };

    bool prove_and_verify(const API::Flags& flags,
                          const std::filesystem::path& bytecode_path,
                          const std::filesystem::path& witness_path) override
    {
        if (!flags.input_type || !(*flags.input_type == "compiletime_stack" || *flags.input_type == "runtime_stack")) {
            throw_or_abort("No input_type or input_type not supported");
        }
        std::vector<acir_format::AcirProgram> folding_stack =
            _build_folding_stack(*flags.input_type, bytecode_path, witness_path);
        ClientIVC ivc = _accumulate(folding_stack);
        const bool verified = ivc.prove_and_verify();
        return verified;
    };

    void gates([[maybe_unused]] const API::Flags& flags,
               [[maybe_unused]] const std::filesystem::path& bytecode_path,
               [[maybe_unused]] const std::filesystem::path& witness_path) override
    {
        throw_or_abort("API function not implemented");
    };

    void contract([[maybe_unused]] const API::Flags& flags,
                  [[maybe_unused]] const std::filesystem::path& output_path,
                  [[maybe_unused]] const std::filesystem::path& vk_path) override
    {
        throw_or_abort("API function not implemented");
    };

    void to_fields([[maybe_unused]] const API::Flags& flags,
                   [[maybe_unused]] const std::filesystem::path& proof_path,
                   [[maybe_unused]] const std::filesystem::path& vk_path,
                   [[maybe_unused]] const std::filesystem::path& output_path) override
    {
        throw_or_abort("API function not implemented");
    };
};
} // namespace bb
