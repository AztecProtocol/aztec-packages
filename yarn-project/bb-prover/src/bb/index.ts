#!/usr/bin/env -S node --no-warnings
import { createConsoleLogger } from '@aztec/foundation/log';

import 'source-map-support/register.js';

import { getProgram } from './cli.js';

const log = createConsoleLogger();

/** CLI main entrypoint */
async function main() {
  process.once('SIGINT', () => process.exit(0));
  process.once('SIGTERM', () => process.exit(0));

  const program = getProgram(log);
  await program.parseAsync(process.argv);
}

main().catch(err => {
  log(`Error in command execution`);
  log(err);
  process.exit(1);
});
