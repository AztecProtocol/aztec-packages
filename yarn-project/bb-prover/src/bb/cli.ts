import { type LogFn } from '@aztec/foundation/log';
import {
  type ProtocolArtifact,
  ProtocolCircuitArtifacts,
  type ServerProtocolArtifact,
} from '@aztec/noir-protocol-circuits-types';

import { Command } from 'commander';
import * as fs from 'fs/promises';

import { generateContractForCircuit, generateKeyForNoirCircuit } from './execute.js';

const { BB_WORKING_DIRECTORY, BB_BINARY_PATH } = process.env;

/**
 * Returns commander program that defines the CLI.
 * @param log - Console logger.
 * @returns The CLI.
 */
export function getProgram(log: LogFn): Command {
  const program = new Command('bb-cli');

  program.description('CLI for interacting with Barretenberg.');

  program
    .command('protocol-circuits')
    .description('Lists the available protocol circuit artifacts')
    .action(() => {
      log(Object.keys(ProtocolCircuitArtifacts).reduce((prev: string, x: string) => prev.concat(`\n${x}`)));
    });

  program
    .command('write-vk')
    .description('Generates the verification key for the specified circuit')
    .requiredOption(
      '-w, --working-directory <string>',
      'A directory to use for storing input/output files',
      BB_WORKING_DIRECTORY,
    )
    .requiredOption('-b, --bb-path <string>', 'The path to the BB binary', BB_BINARY_PATH)
    .requiredOption('-c, --circuit <string>', 'The name of a protocol circuit')
    .requiredOption('-f, --flavor <string>', 'The name of the verification key flavor', 'ultra_honk')
    .action(async options => {
      const compiledCircuit = ProtocolCircuitArtifacts[options.circuit as ProtocolArtifact];
      if (!compiledCircuit) {
        log(`Failed to find circuit ${options.circuit}`);
        return;
      }
      try {
        await fs.access(options.workingDirectory, fs.constants.W_OK);
      } catch (error) {
        log(`Working directory does not exist`);
        return;
      }
      await generateKeyForNoirCircuit(
        options.bbPath,
        options.workingDirectory,
        options.circuit,
        compiledCircuit,
        options.flavor,
        // (options.circuit as ServerProtocolArtifact) === 'RootRollupArtifact' ? 'ultra_keccak_honk' : 'ultra_honk',
        log,
      );
    });

  program
    .command('write-contract')
    .description('Generates the verification contract for the specified circuit')
    .requiredOption(
      '-w, --working-directory <string>',
      'A directory to use for storing input/output files',
      BB_WORKING_DIRECTORY,
    )
    .requiredOption('-b, --bb-path <string>', 'The path to the BB binary', BB_BINARY_PATH)
    .requiredOption('-c, --circuit <string>', 'The name of a protocol circuit')
    .requiredOption('-n --contract-name <string>', 'The name of the contract to generate', 'contract.sol')
    .action(async options => {
      const compiledCircuit = ProtocolCircuitArtifacts[options.circuit as ProtocolArtifact];
      if (!compiledCircuit) {
        log(`Failed to find circuit ${options.circuit}`);
        return;
      }
      try {
        await fs.access(options.workingDirectory, fs.constants.W_OK);
      } catch (error) {
        log(`Working directory does not exist`);
        return;
      }

      await generateContractForCircuit(
        options.bbPath,
        options.workingDirectory,
        options.circuit,
        compiledCircuit,
        options.contractName,
        log,
      );
    });

  return program;
}
