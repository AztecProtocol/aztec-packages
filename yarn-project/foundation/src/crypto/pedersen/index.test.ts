import { BarretenbergSync } from '@aztec/bb.js';

import { toBufferBE } from '../../bigint-buffer/index.js';
import { pedersenCommit, pedersenHash, pedersenHashBuffer } from './index.js';

describe('pedersen', () => {
  beforeAll(async () => {
    await BarretenbergSync.initSingleton();
  });

  it('pedersen commit', () => {
    const r = pedersenCommit([toBufferBE(1n, 32), toBufferBE(1n, 32)]);
    expect(r).toEqual([
      Buffer.from('2f7a8f9a6c96926682205fb73ee43215bf13523c19d7afe36f12760266cdfe15', 'hex'),
      Buffer.from('01916b316adbbf0e10e39b18c1d24b33ec84b46daddf72f43878bcc92b6057e6', 'hex'),
    ]);
  });

  it('pedersen commit with zero', () => {
    const r = pedersenCommit([toBufferBE(0n, 32), toBufferBE(1n, 32)]);
    expect(r).toEqual([
      Buffer.from('054aa86a73cb8a34525e5bbed6e43ba1198e860f5f3950268f71df4591bde402', 'hex'),
      Buffer.from('209dcfbf2cfb57f9f6046f44d71ac6faf87254afc7407c04eb621a6287cac126', 'hex'),
    ]);
  });

  it('pedersen hash', () => {
    const r = pedersenHash([toBufferBE(1n, 32), toBufferBE(1n, 32)]);
    expect(r.toString()).toEqual('0x07ebfbf4df29888c6cd6dca13d4bb9d1a923013ddbbcbdc3378ab8845463297b');
  });

  it('pedersen hash with index', () => {
    const r = pedersenHash([toBufferBE(1n, 32), toBufferBE(1n, 32)], 5);
    expect(r.toString()).toEqual('0x1c446df60816b897cda124524e6b03f36df0cec333fad87617aab70d7861daa6');
  });

  it('pedersen hash buffer', () => {
    const input = Buffer.alloc(123);
    input.writeUint32BE(321, 0);
    input.writeUint32BE(456, 119);
    const r = pedersenHashBuffer(input);
    expect(r).toMatchSnapshot();
  });
});
