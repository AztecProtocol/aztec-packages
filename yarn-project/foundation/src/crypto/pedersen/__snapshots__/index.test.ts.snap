// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pedersen pedersen hash buffer 1`] = `
{
  "data": [
    43,
    213,
    196,
    82,
    160,
    201,
    113,
    98,
    41,
    79,
    201,
    223,
    208,
    241,
    224,
    157,
    14,
    9,
    201,
    95,
    165,
    237,
    63,
    241,
    73,
    251,
    222,
    243,
    102,
    203,
    81,
    249,
  ],
  "type": "Buffer",
}
`;
